### PR TITLE
Only include log structures if enabled

### PR DESCRIPTION
**Pull request by IamPete1**
A few more savings like https://github.com/ArduPilot/ardupilot/pull/19678.

### CODE REVIEW

**Review by andyp1per — APPROVED** [review]
Nice!

File: libraries/AP_VisualOdom/LogStructure.h [hunk 1]
```diff
@@ -87,11 +88,14 @@ struct PACKED log_VisualVelocity {
     uint8_t ignored;
 };
 
+#if !HAL_VISUALODOM_ENABLED
```
Comment (amilcarlucas) [inline libraries/AP_VisualOdom/LogStructure.h:91]: I like the pattern without the ! better

**Review by amilcarlucas — CHANGES_REQUESTED** [review]

File: libraries/AP_VisualOdom/LogStructure.h (hunk 1 above)
Comment (IamPete1) [reply libraries/AP_VisualOdom/LogStructure.h]: done

**Comment by IamPete1** [comment]
Dropped the ESC telem change due to build issues.

**Comment by IamPete1** [comment]
Costs flash on 2M boards, due to the slight change in includes I think.
```
Binary Name      Text [B]       Data [B]     BSS (B)        Total Flash Change [B] (%)      Flash Free After PR (B)
---------------  -------------  -----------  -------------  ----------------------------  -------------------------
arduplane        36 (+0.0020%)  0 (0.0000%)  4 (+0.0015%)   36 (+0.0020%)                                    206104
ardurover        36 (+0.0022%)  0 (0.0000%)  4 (+0.0015%)   36 (+0.0022%)                                    362928
arducopter-heli  36 (+0.0020%)  0 (0.0000%)  -4 (-0.0015%)  36 (+0.0020%)                                    180376
arducopter       36 (+0.0020%)  0 (0.0000%)  4 (+0.0015%)   36 (+0.0020%)                                    188248
antennatracker   36 (+0.0026%)  0 (0.0000%)  4 (+0.0015%)   36 (+0.0026%)                                    604256
blimp            36 (+0.0028%)  0 (0.0000%)  4 (+0.0015%)   36 (+0.0028%)                                    663428
ardusub          36 (+0.0023%)  0 (0.0000%)  -4 (-0.0015%)  36 (+0.0023%)                                    402156
```
Saves on 1M boards.
```
Binary Name      Text [B]         Data [B]     BSS (B)        Total Flash Change [B] (%)      Flash Free After PR (B)
---------------  ---------------  -----------  -------------  ----------------------------  -------------------------
arduplane        -564 (-0.0580%)  0 (0.0000%)  4 (+0.0031%)   -564 (-0.0579%)                                    9920
ardurover        -568 (-0.0663%)  0 (0.0000%)  0 (0.0000%)    -568 (-0.0663%)                                  126280
arducopter-heli  -556 (-0.0589%)  0 (0.0000%)  -4 (-0.0031%)  -556 (-0.0588%)                                   37688
arducopter       -552 (-0.0591%)  0 (0.0000%)  0 (0.0000%)    -552 (-0.0590%)                                   47952
antennatracker   -564 (-0.0851%)  0 (0.0000%)  4 (+0.0031%)   -564 (-0.0850%)                                  320056
blimp            -568 (-0.0834%)  0 (0.0000%)  0 (0.0000%)    -568 (-0.0833%)                                  302092
ardusub          -564 (-0.0694%)  0 (0.0000%)  4 (+0.0031%)   -564 (-0.0693%)                                  169828
```

File: libraries/AP_VisualOdom/LogStructure.h (hunk 1 above)
Comment (IamPete1) [reply libraries/AP_VisualOdom/LogStructure.h]: done

File: libraries/AP_Proximity/LogStructure.h [hunk 2]
```diff
@@ -1,6 +1,7 @@
 #pragma once
 
 #include <AP_Logger/LogStructure.h>
+#include "AP_Proximity.h"
```
Comment (peterbarker) [inline libraries/AP_Proximity/LogStructure.h:4]: The problem with doing things this way is that now every header file in ArduPilot will depend on every other header in ArduPilot, as AP_Logger/LogStructure is commonly used.  I can also see us getting into include-loop-hell (you've probably noticed we skirt that here already!

I like the idea of excluding the log structures like this (can't argue with those results!), but we could instead consider moving `HAL_PROXIMITY_ENABLED` out of `AP_Proximity.h` and into `AP_HAL/AP_HAL_Boards.h`.  Making this move came up for other reasons on a DevCall the other week....

Comment (IamPete1) [reply libraries/AP_Proximity/LogStructure.h]: Moved to `...config.h` files

**Comment by IamPete1** [comment]
I had another bash at ESC telem with the new `config.h` approach. Still not luck in getting it to build without starting down a big rabbit hole. I suspect we have something not quite right with how the defines are implemented for that library.

**Comment by IamPete1** [comment]
Saves on 1M boards:
```
Binary Name      Text [B]         Data [B]     BSS (B)        Total Flash Change [B] (%)      Flash Free After PR (B)
---------------  ---------------  -----------  -------------  ----------------------------  -------------------------
arducopter-heli  -556 (-0.0550%)  0 (0.0000%)  -4 (-0.0020%)  -556 (-0.0550%)                                   21100
arduplane        -568 (-0.0553%)  0 (0.0000%)  0 (0.0000%)    -568 (-0.0552%)                                    4532
arducopter       -552 (-0.0553%)  0 (0.0000%)  0 (0.0000%)    -552 (-0.0552%)                                   33628
ardurover        -564 (-0.0615%)  0 (0.0000%)  4 (+0.0020%)   -564 (-0.0614%)                                  114844
blimp            -568 (-0.0795%)  0 (0.0000%)  0 (0.0000%)    -568 (-0.0794%)                                  317716
ardusub          -564 (-0.0614%)  0 (0.0000%)  4 (+0.0020%)   -564 (-0.0613%)                                  113416
antennatracker   -564 (-0.0745%)  0 (0.0000%)  4 (+0.0020%)   -564 (-0.0744%)                                  274352
```

No change on 2M as they have all the features enabled anyway:
```
Binary Name      Text [B]     Data [B]     BSS (B)      Total Flash Change [B] (%)      Flash Free After PR (B)
---------------  -----------  -----------  -----------  ----------------------------  -------------------------
arducopter-heli  0 (0.0000%)  0 (0.0000%)  0 (0.0000%)  0 (0.0000%)                                      179560
arduplane        0 (0.0000%)  0 (0.0000%)  0 (0.0000%)  0 (0.0000%)                                      205232
arducopter       0 (0.0000%)  0 (0.0000%)  0 (0.0000%)  0 (0.0000%)                                      187504
ardurover        0 (0.0000%)  0 (0.0000%)  0 (0.0000%)  0 (0.0000%)                                      362424
blimp            0 (0.0000%)  0 (0.0000%)  0 (0.0000%)  0 (0.0000%)                                      662796
ardusub          0 (0.0000%)  0 (0.0000%)  0 (0.0000%)  0 (0.0000%)                                      401644
antennatracker   0 (0.0000%)  0 (0.0000%)  0 (0.0000%)  0 (0.0000%)                                      603552
```

**Review by tridge — APPROVED** [review]

**Comment by amilcarlucas** [comment]
There are multiple ESC device drivers that depend on the ESC Telem, I bet that is causing some of the issues you saw.

**Pull request merged**